pkg/storageops/vsphere/vsphere.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		logrus.Errorf("Failed to attach vsphere disk: %s for VM: %s. err: +%v", diskPath, vmObj.Name(), err)
+		return "", err
+	}
+
+	return path.Join(diskByIDPath, diskSCSIPrefix+diskUUID), nil
+}
+
+func (ops *vsphereOps) Detach(diskPath string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	vmObj, err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return err
+	}
+
+	if err := vmObj.DetachDisk(ctx, diskPath); err != nil {
+		logrus.Errorf("Failed to detach vsphere disk: %s for VM: %s. err: +%v", diskPath, vmObj.Name(), err)
+		return err
+	}
+
+	return nil
+}
+
+// Delete virtual disk at given path
+func (ops *vsphereOps) Delete(diskPath string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	vmObj, err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return err
+	}
+
+	disk := diskmanagers.VirtualDisk{
+		DiskPath:      diskPath,
+		VolumeOptions: &vclib.VolumeOptions{},
+		VMOptions:     &vclib.VMOptions{},
+	}
+
+	err = disk.Delete(ctx, vmObj.Datacenter)
+	if err != nil {
+		logrus.Errorf("Failed to delete vsphere disk: %s. err: %+v", diskPath, err)
+	}
+
+	return err
+}
+
+// Desribe an instance of the virtual machine object to which ops is connected to
+func (ops *vsphereOps) Describe() (interface{}, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	return ops.renewVM(ctx, ops.vm)
+}
+
+// FreeDevices is not supported by this provider
+func (ops *vsphereOps) FreeDevices(blockDeviceMappings []interface{}, rootDeviceName string) ([]string, error) {
+	return nil, storageops.ErrNotSupported
+}
+
+func (ops *vsphereOps) Inspect(diskPaths []*string) ([]interface{}, error) {
+	// TODO find a way to map diskPaths to unattached/attached virtual disks and query info
+	// currently returning the disks directly
+
+	return nil, storageops.ErrNotSupported
+}
+
+// DeviceMappings returns map[local_attached_volume_path]->volume ID/NAME
+func (ops *vsphereOps) DeviceMappings() (map[string]string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	vmObj, err := ops.renewVM(ctx, ops.vm)
+	if err != nil {
+		return nil, err
+	}
+
+	vmDevices, err := vmObj.Device(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get devices for vm: %s", vmObj.Name())
+	}
+
+	// Go over all the devices attached on this vm and create a map of just the virtual disks and where
+	// they are attached on the vm
+	m := make(map[string]string)
+	for _, device := range vmDevices {
+		if vmDevices.TypeName(device) == "VirtualDisk" {
+			virtualDevice := device.GetVirtualDevice()
+			backing, ok := virtualDevice.Backing.(*types.VirtualDiskFlatVer2BackingInfo)
+			if ok {
+				devicePath, err := ops.DevicePath(backing.FileName)
+				if err == nil && len(devicePath) != 0 { // TODO can ignore errors?
```

Comment:
nit: we are calling renewVM twice, once in this function and once in DevicePath.

Reply:
That's because `DevicePath` is also an interface function by itself and hence needs to ensure it has a renewed vm. With the updated renewVM implementation now, it will be less expensive.